### ISSUE TITLE
Issue: free(): invalid pointer Aborted (core dumped)

### ISSUE THREAD

**Issue by fei67896** [Asatheesh6561/MAPPO-LCE#2]
Hi, thank you for you share! The work is excellent!
When I run the repo with wandb, I met a problem：

"new test result : -850.5623309999996
free(): invalid pointer
Aborted (core dumped)"

How can I fix this issue, thanks a lot!

##################Hardware
PyTorch  2.5.1
Python  3.12(ubuntu22.04)
CUDA  12.4
GPU RTX 3090(24GB) * 1
CPU 15 vCPU Intel(R) Xeon(R) Platinum 8358P CPU @ 2.60GHz
RAM 90GB
Hard disk :30 GB
##################conda env:
_libgcc_mutex             0.1                        main    
_openmp_mutex             5.1                       1_gnu    
annotated-types           0.7.0                    pypi_0    pypi
ca-certificates           2025.4.26            hbd8a1cb_0    conda-forge
certifi                   2025.4.26                pypi_0    pypi
charset-normalizer        3.4.2                    pypi_0    pypi
cityflow                  0.1                      pypi_0    pypi
click                     8.1.8                    pypi_0    pypi
cloudpickle               3.1.1                    pypi_0    pypi
colorama                  0.4.6                    pypi_0    pypi
dataclasses               0.6                      pypi_0    pypi
docker-pycreds            0.4.0                    pypi_0    pypi
docopt-ng                 0.9.0                    pypi_0    pypi
eval-type-backport        0.2.2                    pypi_0    pypi
filelock                  3.18.0                   pypi_0    pypi
fsspec                    2025.5.1                 pypi_0    pypi
gitdb                     4.0.12                   pypi_0    pypi
gitpython                 3.1.44                   pypi_0    pypi
gym                       0.26.2                   pypi_0    pypi
gym-notices               0.0.8                    pypi_0    pypi
idna                      3.10                     pypi_0    pypi
importlib-metadata        8.7.0                    pypi_0    pypi
jinja2                    3.1.6                    pypi_0    pypi
jsonpickle                4.1.0                    pypi_0    pypi
ld_impl_linux-64          2.40                 h12ee557_0    
libffi                    3.4.4                h6a678d5_1    
libgcc                    15.1.0               h767d61c_2    conda-forge
libgcc-ng                 15.1.0               h69a702a_2    conda-forge
libgomp                   15.1.0               h767d61c_2    conda-forge
libstdcxx                 15.1.0               h8f9b012_2    conda-forge
libstdcxx-ng              15.1.0               h4852527_2    conda-forge
markupsafe                3.0.2                    pypi_0    pypi
mpmath                    1.3.0                    pypi_0    pypi
munch                     4.0.0                    pypi_0    pypi
ncurses                   6.4                  h6a678d5_0    
networkx                  3.2.1                    pypi_0    pypi
numpy                     2.0.2                    pypi_0    pypi
nvidia-cublas-cu12        12.6.4.1                 pypi_0    pypi
nvidia-cuda-cupti-cu12    12.6.80                  pypi_0    pypi
nvidia-cuda-nvrtc-cu12    12.6.77                  pypi_0    pypi
nvidia-cuda-runtime-cu12  12.6.77                  pypi_0    pypi
nvidia-cudnn-cu12         9.5.1.17                 pypi_0    pypi
nvidia-cufft-cu12         11.3.0.4                 pypi_0    pypi
nvidia-cufile-cu12        1.11.1.6                 pypi_0    pypi
nvidia-curand-cu12        10.3.7.77                pypi_0    pypi
nvidia-cusolver-cu12      11.7.1.2                 pypi_0    pypi
nvidia-cusparse-cu12      12.5.4.2                 pypi_0    pypi
nvidia-cusparselt-cu12    0.6.3                    pypi_0    pypi
nvidia-nccl-cu12          2.26.2                   pypi_0    pypi
nvidia-nvjitlink-cu12     12.6.85                  pypi_0    pypi
nvidia-nvtx-cu12          12.6.77                  pypi_0    pypi
openssl                   3.5.0                h7b32b05_1    conda-forge
packaging                 25.0                     pypi_0    pypi
pandas                    2.2.3                    pypi_0    pypi
pillow                    11.2.1                   pypi_0    pypi
pip                       25.1               pyhc872135_2    
platformdirs              4.3.8                    pypi_0    pypi
protobuf                  6.31.0                   pypi_0    pypi
psutil                    7.0.0                    pypi_0    pypi
py-cpuinfo                9.0.0                    pypi_0    pypi
pydantic                  2.11.5                   pypi_0    pypi
pydantic-core             2.33.2                   pypi_0    pypi
python                    3.9.21               he870216_1    
python-dateutil           2.9.0.post0              pypi_0    pypi
pytz                      2025.2                   pypi_0    pypi
pyyaml                    6.0.2                    pypi_0    pypi
pyyaml-include            2.2                      pypi_0    pypi
readline                  8.2                  h5eee18b_0    
requests                  2.32.3                   pypi_0    pypi
sacred                    0.8.7                    pypi_0    pypi
sentry-sdk                2.29.1                   pypi_0    pypi
setproctitle              1.3.6                    pypi_0    pypi
setuptools                78.1.1           py39h06a4308_0    
six                       1.17.0                   pypi_0    pypi
smmap                     5.0.2                    pypi_0    pypi
sqlite                    3.45.3               h5eee18b_0    
sympy                     1.14.0                   pypi_0    pypi
tabulate                  0.9.0                    pypi_0    pypi
tensorboardx              2.6.2.2                  pypi_0    pypi
tk                        8.6.14               h39e8969_0    
torch                     2.7.0                    pypi_0    pypi
torchvision               0.22.0                   pypi_0    pypi
tqdm                      4.67.1                   pypi_0    pypi
triton                    3.3.0                    pypi_0    pypi
typing-extensions         4.13.2                   pypi_0    pypi
typing-inspection         0.4.1                    pypi_0    pypi
tzdata                    2025.2                   pypi_0    pypi
urllib3                   2.4.0                    pypi_0    pypi
wandb                     0.19.11                  pypi_0    pypi
wheel                     0.45.1           py39h06a4308_0    
wrapt                     1.17.2                   pypi_0    pypi
xz                        5.6.4                h5eee18b_1    
zipp                      3.21.0                   pypi_0    pypi
zlib                      1.2.13               h5eee18b_1  

##################log
[INFO 17:53:15] main Beginning training for 500000 timesteps
[INFO 17:55:55] main t_env: 5040 / 500000
[INFO 17:55:55] main Estimated time left: 4 hours, 21 minutes, 45 seconds. Time passed: 2 minutes, 39 seconds
new test result : -1122.3929484999996
[INFO 17:58:59] main t_env: 10080 / 500000
[INFO 17:58:59] main Estimated time left: 4 hours, 58 minutes, 19 seconds. Time passed: 5 minutes, 44 seconds
new test result : -1097.5326975
[INFO 17:59:15] main Recent Stats | t_env:      10080 | Episode:      224

[INFO 18:02:06] main t_env: 15120 / 500000
[INFO 18:02:06] main Estimated time left: 4 hours, 59 minutes, 48 seconds. Time passed: 8 minutes, 51 seconds
new test result : -850.5623309999996
free(): invalid pointer
Aborted (core dumped)

**Comment by Asatheesh6561** [comment]
Hi fei67896,

Thank you for being interested in our work. Unfortunately this error stems from the general instability of the Cityflow C++ environment, as referenced by https://github.com/zyr17/CityFlow. We tried moving some computation of stats to Python instead of C++, but this made the run time much slower (5 hours -> 9 days estimated). 

Maybe you can try running again? We had similar issues and sometimes it would be fixed by re-running. However since it is a low level memory error we are unable to even catch the error, let alone fix it. If you find a fix for this please submit a pull request and let us know.

We apologize for this inconvenience, and we once again thank you for your interest in our paper.

**Comment by fei67896** [comment]
Thank you for your reply.
Is there any method of resume from breakpoint in the repo? I met this issue twice at almost finish the training, so it is disappointing.  Or any trick please?

**Comment by Asatheesh6561** [comment]
You can use the save and load model functions in the learner. We chose not to save the model as they can take up a lot of memory.

**Comment by Asatheesh6561** [comment]
I'm closing this issue, as it seems to be resolved. If there are any other problems, please feel free to open this issue again.
